src/expr-derive-impl/src/sqlfunc.rs

```diff
@@ -0,0 +1,441 @@
+            if let syn::Type::Reference(r) = &*pat.ty {
+                if r.lifetime.is_none() {
+                    let ty = syn::Type::Reference(syn::TypeReference {
+                        lifetime: Some(Lifetime::new("'a", r.span())),
+                        ..r.clone()
+                    });
+                    return Ok(ty);
+                }
+            }
+            Ok((*pat.ty).clone())
+        }
+        _ => Err(syn::Error::new(
+            arg.sig.inputs[nth].span(),
+            "Unsupported argument type",
+        )),
+    }
+}
+
+/// Determine the output type for a function. Returns an error if the function
+/// does not return a value.
+fn output_type(arg: &syn::ItemFn) -> Result<&syn::Type, syn::Error> {
+    match &arg.sig.output {
+        syn::ReturnType::Type(_, ty) => Ok(&*ty),
+        syn::ReturnType::Default => Err(syn::Error::new(
+            arg.sig.output.span(),
+            "Function needs to return a value",
+        )),
+    }
+}
+
+/// Produce a `EagerUnaryFunc` implementation.
+fn unary_func(
+    func: &syn::ItemFn,
+    modifiers: Modifiers,
+    arena: bool,
+) -> darling::Result<TokenStream> {
+    let fn_name = &func.sig.ident;
+    let struct_name = camel_case(&func.sig.ident);
+    let input_ty = arg_type(func, 0)?;
+    let output_ty = output_type(func)?;
+    let Modifiers {
+        is_monotone,
+        sqlname,
+        preserves_uniqueness,
+        inverse,
+        is_infix_op,
+        output_type,
+        negate,
+        could_error,
+        propagates_nulls,
+    } = modifiers;
+
+    if is_infix_op.is_some() {
+        return Err(darling::Error::unknown_field("is_infix_op"));
+    }
+    if negate.is_some() {
+        return Err(darling::Error::unknown_field("negate"));
+    }
+    if propagates_nulls.is_some() {
+        return Err(darling::Error::unknown_field("propagates_nulls"));
+    }
+
+    let preserves_uniqueness_fn = preserves_uniqueness.map(|preserves_uniqueness| {
+        quote! {
+            fn preserves_uniqueness(&self) -> bool {
+                #preserves_uniqueness
+            }
+        }
+    });
+
+    let inverse_fn = inverse.as_ref().map(|inverse| {
+        quote! {
+            fn inverse(&self) -> Option<crate::UnaryFunc> {
+                #inverse
+            }
+        }
+    });
+
+    let is_monotone_fn = is_monotone.map(|is_monotone| {
+        quote! {
+            fn is_monotone(&self) -> bool {
+                #is_monotone
+            }
+        }
+    });
+
+    let name = if let Some(sqlname) = sqlname {
+        quote! {
+            #sqlname
+        }
+    } else {
+        quote! {
+            stringify!(#fn_name)
+        }
+    };
+
+    let (output_type, introduces_nulls_fn) = if let Some(output_type) = output_type {
+        let introduces_nulls_fn = quote! {
+            fn introduces_nulls(&self) -> bool {
+                <#output_type as mz_repr::DatumType<'_, ()>>::nullable()
```

Comment:
in a macro when referencing an external crate I think it's best practice to prefix it with a `::`, e.g.
```
<#output_type as ::mz_repr::DatumType<'_, ()>>::nullable()
```
to ensure it doesn't conflict with a module of the same name in the crate the macro is being used. Not a huge deal, but might help someone in the future, especially for a crate like `serde` which _maybe_ could also be the name of a module

Reply:
Good point, fixed.